### PR TITLE
[NT-1851] Enable/Disable Segment By Feature Flag

### PR DESCRIPTION
**Pull request by justinswart**
# 📲 What

- Moves our configuration of Segment to be after the `AppEnvironment` is set up.
- Enables and disables all Segment tracking by feature flag.
- Tidies up our protocols that we were using before to mock and feature flag tracking.

# 🤔 Why

- We had a race condition in that we were instantiating Segment's `Analytics` during the instantiation of the `AppEnvironment` and therefore could not inspect the environment at the same time. Moving this makes more tests possible.
- Previously we were feature-flagging the `track`, `identify`, etc methods but Segment would still track app lifecycle events. It turns out Segment provides `disable()` and `enable()` functions to completely disable/enable tracking which we are now using instead.
  - Note: Segment is _disabled_ by default and enabled by the feature flag configuration.
- We no longer need some of the protocol functions as we don't need to test the feature flagging at those particular call-sites.

# 🛠 How

- Moved Segment initialization to be driven by an `AppDelegateViewModel` output.
- Observing configuration changes and enabling/disabling Segment.
- Tidied up these protocols.

# ✅ Acceptance criteria

Open Segment's debugger and run the app.

- [ ] Toggle the Segment feature flag in the app's debug menu, observe that events are or are not sent to Segment based on the state of the flag.
- [ ] Trigger a reload of the config and intercept it using Charles Proxy. Edit the response to enable or disable Segment using the feature flag. Observe that events are or are not sent to Segment based on the state of the flag.

### CODE REVIEW

File: Kickstarter-iOS/AppDelegate.swift [hunk 1]
```diff
@@ -263,6 +263,30 @@ internal final class AppDelegate: UIResponder, UIApplicationDelegate {
...
+    self.viewModel.outputs.segmentIsEnabled
+      .observeValues { enabled in
+        if enabled {
+          Analytics.shared().enable()
+        } else {
+          Analytics.shared().disable()
+        }
+      }
+
+    NotificationCenter.default
```
Comment (justinswart) [inline Kickstarter-iOS/AppDelegate.swift:281]: Ordinarily we would hang on to the token for these observers and release them but the `AppDelegate` is never released so I guess it's ok here 😄

File: Library/Tracking/Segment.swift [hunk 2]
```diff
@@ -7,54 +7,27 @@ public extension Analytics {
...
  The `TrackingClientType` and `IdentifyingTrackingClient` protocols
- allow us to create mocks to test these code paths. The protocol exposes functions that are named similarly
- but differently so that we can perform the check to see that tracking is enabled before calling any of the
- functions on the library itself so as to not unintentionally contribute to tracking data during debugging.
+ allow us to create mocks to test these code paths.
  */
-
-extension Analytics: IdentifyingTrackingClient {
-  public func identify(userId: String?, traits: [String: Any]?) {
-    guard AppEnvironment.current.environmentVariables.isTrackingEnabled,
```
Comment (justinswart) [inline Library/Tracking/Segment.swift:39]: Weirdly no tests failed removing this so it looks like we weren't testing it anyway 🤦‍♂️

File: Configs/Secrets.swift.example [hunk 3]
```diff
@@ -40,8 +40,8 @@ public enum Secrets {
   }
 
   public enum Segment {
-    public static let staging = "write-key"
-    public static let production = "write-key"
+    public static let staging = "write-key-production"
+    public static let production = "write-key-staging"
```
Comment (singhhari) [inline Configs/Secrets.swift.example:44]: looks like you mixed these up

File: Kickstarter-iOS/AppDelegate.swift [hunk 4]
```diff
@@ -263,6 +263,30 @@ internal final class AppDelegate: UIResponder, UIApplicationDelegate {
...
 
+    self.viewModel.outputs.configureSegment
+      .observeValues { writeKey in
+        let client = Analytics.configuredClient(withWriteKey: writeKey)
+        AppEnvironment.current.ksrAnalytics.configureSegmentClient(client)
+      }
+
+    self.viewModel.outputs.segmentIsEnabled
+      .observeValues { enabled in
+        if enabled {
```
Comment (singhhari) [inline Kickstarter-iOS/AppDelegate.swift:274]: would a ternary operator work here?

File: Configs/Secrets.swift.example (hunk 3 above)
Comment (justinswart) [reply Configs/Secrets.swift.example:44]: oh lol yeah, they mostly just need to not be the same but best have them the right way around 👍

File: Kickstarter-iOS/AppDelegate.swift (hunk 4 above)
Comment (justinswart) [reply Kickstarter-iOS/AppDelegate.swift:274]: Don't see why not! 👍

**Pull request merged**